state.go

```diff
@@ -1307,11 +1313,19 @@ func StoreStateData(store store.Store, sd StateData) error {
 	}
 	data, _ := json.Marshal(sd)
 
+	return store.WriteAll(stateDataKeyCustom, data)
+}
+
+func CommitStateData(store store.Store) error {
+	data, err := store.ReadAll(stateDataKeyCustom)
+	if err != nil {
+		return err
+	}
 	return store.WriteAll(stateDataKey, data)
```

Comment:
The read and write should be done in one DB transaction to avoid race conditions. It's true there shouldn't be other writers to the database at this time, but who knows what will happen in the future, and I'd rather get our DB handling correct from the start instead of dealing with the long term consequences like in CFEngine.

Unfortunately our DB layer is not good at multi-operation transactions, even though LMDB supports it, so we may have to extend the interface a little bit.

Reply:
Actually, I think I can implement the transaction interface, since I need it for something else right now. Then you can use it afterwards.